fq.go

```diff
@@ -0,0 +1,163 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+)
+
+type FQScheduler struct {
+	lock   *sync.Mutex
+	queues []*Queue
+	vt     uint64
+	C      uint64
+	G      uint64
+	seen   bool
+}
+
+// TODO(aaron-prindle) add concurrency enforcement - 'C'
+
+func (q *FQScheduler) chooseQueue(packet *Packet) *Queue {
+	for _, queue := range q.queues {
+		if packet.key == queue.key {
+			return queue
+		}
+	}
+	panic("no matching queue for packet")
+}
+
+func newFQScheduler(queues []*Queue) *FQScheduler {
+	now := NowAsUnixMilli()
+	fq := &FQScheduler{
+		lock:   &sync.Mutex{},
+		queues: queues,
+		// R(t) = (server start time) + (1 ns) * (number of rounds since server start).
+		vt: now,
```

Comment:
The baseline does not really matter, it could just as well be zero (and that would give you more significant bits to work with).

Reply:
Ah makes sense, thanks.